### PR TITLE
[WIP]: Basic support for aws_lambda_function publish attribute

### PR DESCRIPTION
**Pull request by kmamykin**
Looking for review.

Not ready to merge yet. 

* [ ] test with publish = false for backward compatibility
* [ ] test with aliases
* [ ] test with permissions
* [ ] test with mappings

### CODE REVIEW

File: builtin/providers/aws/resource_aws_lambda_function.go [hunk 1]
```diff
@@ -356,6 +342,45 @@ func resourceAwsLambdaFunctionUpdate(d *schema.ResourceData, meta interface{}) e
...
+	}
+
+	log.Printf("[DEBUG] Send Update Lambda Function Code request: %#v", codeReq)
+	if codeUpdate {
+		updateFunctionCodeOutput, err := conn.UpdateFunctionCode(codeReq)
+		if err != nil {
+			return fmt.Errorf("Error modifying Lambda Function Code %s: %s", d.Id(), err)
+		}
+
+		d.Set("version", updateFunctionCodeOutput.Version)
```
Comment (kmamykin) [inline builtin/providers/aws/resource_aws_lambda_function.go]: Do I need to SetPartial for version here as well?

Comment (catsby) [inline builtin/providers/aws/resource_aws_lambda_function.go]: Looks right to me

**Comment by catsby** [comment]
Thanks @kmamykin – I edited the title to note the WIP status, just remove that and ping here when you're ready! Let us know if you have any questions

**Comment by sinangedik** [comment]
Could this be merged in?

**Comment by kmamykin** [comment]
Well, the publish=true is working. However there is still a problem: When a new function is deployed with publish=true, the resource is updated with a new version and ARN (calculated attributes), but any dependent resources that depend on function's ARN are not updated. `terraform plan` shows that the function itself with new source_code will be deployed, but no updates calculated for dependent resource.

That looks like its a more general problem with terraform rather then an issue with this PR. But I think merging this PR as will confuse people. 

Any idea how to control dependencies on calculated properties?

When I get a bit more time, I will post cmd output.

**Comment by radeksimko** [comment]
> That looks like its a more general problem with terraform rather then an issue with this PR.

Agreed.

> But I think merging this PR as is will confuse people.

Agreed.

> Any idea how to control dependencies on calculated attributes?

@kmamykin Unfortunately I don't think there's any way to do that until `CalculatedWhen` becomes a real thing.

See https://github.com/hashicorp/terraform/pull/4846 and https://github.com/hashicorp/terraform/pull/5330

**Comment by radeksimko** [comment]
> But I think merging this PR as is will confuse people.

It would if we merged the current implementation, but I think there's a chance of using local file instead of S3 for the test purposes - that way we can get around the issue and it shouldn't confuse people too much to prevent this PR from merging.

That said the confusing behaviour w/ S3 should probably be documented.

**Comment by radeksimko** [comment]
Hi @kmamykin 
I almost forgot about your PR 😱  and submitted https://github.com/hashicorp/terraform/pull/8653 Sorry!

---

As you mentioned in the code all published Lambda versions are immutable. Changes will always be applied to the latest version.

You're setting `Qualifier` in your changes to `GetFunctionInput` which (I assume) is because you want to get & reset the ARN to a qualified one - otherwise the response is pretty much same as response from `GetFunctionInput` without `Qualifier`.

^ The (potentially negative) side effect is that if someone deployed a new version outside of terraform it would remain unnoticed which is contrary to how most other TF resources behave today.

In addition to that qualified ARNs and unqualified ones aren't interchangeable even if they point to the same version/qualifier. Unqualified ARNs in policies are only valid for invokes via unqualified ARNs. See http://docs.aws.amazon.com/lambda/latest/dg/API_AddPermission.html

> If the qualifier is not specified, the permission is valid only when requests is made using unqualified function ARN.

This leads me to a conclusion that we should output both qualified and unqualified ARN and note that these are mutually exclusive - i.e. user should use either one or the other.

**Comment by kmamykin** [comment]
@radeksimko totally fine, thank you for picking it up and bringing it home. My PR was more like an experiment to see if I would make it work in my CI workflow.
Cheers.

**Comment by radeksimko** [comment]
#8653 merged and will be included in the next release.

@kmamykin Thanks for the contribution.

**Pull request closed without merging**